sdk/go/pulumi-language-go/main.go

```diff
@@ -337,33 +400,55 @@ func (host *goLanguageHost) Run(ctx context.Context, req *pulumirpc.RunRequest)
 		return nil, errors.Wrap(err, "failed to prepare environment")
 	}
 
-	cmd, err := findProgram(host.binary)
-	if err != nil {
-		return nil, err
+	if os.Getenv("PULUMI_GO_USE_RUN") != "" {
+		// feature flag to enable old behavior and use `go run`
+		gobin, err := executable.FindExecutable("go")
+		if err != nil {
+			return nil, err
+		}
+
+		cwd, err := os.Getwd()
+		if err != nil {
+			return nil, errors.Wrap(err, "unable to get current working directory")
+		}
+
+		cmd := exec.Command(gobin, "run", cwd)
+		if err := execProgramCmd(cmd, env); err != nil {
+			return &pulumirpc.RunResponse{
+				Error: err.Error(),
+			}, nil
+		}
+
+		return &pulumirpc.RunResponse{}, nil
 	}
-	cmd.Env = env
-	cmd.Stdout, cmd.Stderr = os.Stdout, os.Stderr
 
-	var errResult string
-	if err := cmd.Run(); err != nil {
-		if exiterr, ok := err.(*exec.ExitError); ok {
-			// If the program ran, but exited with a non-zero error code.  This will happen often, since user
-			// errors will trigger this.  So, the error message should look as nice as possible.
-			if status, stok := exiterr.Sys().(syscall.WaitStatus); stok {
-				err = errors.Errorf("program exited with non-zero exit code: %d", status.ExitStatus())
-			} else {
-				err = errors.Wrapf(exiterr, "program exited unexpectedly")
-			}
-		} else {
-			// Otherwise, we didn't even get to run the program.  This ought to never happen unless there's
-			// a bug or system condition that prevented us from running the language exec.  Issue a scarier error.
-			err = errors.Wrapf(err, "problem executing program (could not run language executor)")
+	// the user can explicitly opt in to using a binary executable by specifying
+	// runtime.options.binary in the Pulumi.yaml
+	program := host.binary
+	if program == "" {
+		// user did not specify a binary and we will compile and run the binary on-demand
+		logging.V(5).Infof("No prebuilt executable specified, attempting invocation via compilation")
+
+		program, err = compileProgramCwd(req.GetProject())
+		if err != nil {
+			return nil, errors.Wrap(err, "error in compiling Go")
 		}
+		//defer os.RemoveAll(program)
```

Comment:
After checking out "go run" some more, it seems it does delete the program after running. So I think we do want a Remove call.

Reply:
Deleting the program after running negates the raison d'etre of this PR - faster execution benefits listed in https://github.com/pulumi/pulumi/issues/10350

We could ship it as a separate change, but I'm for shipping it in as part of this. It's about risk. The PR compounds the risks of changed exit codes and possible TEMP saturation. But so far both appear quite tolerable. Shall we just ship as is?

Reply:
You still get other benefits for measurements and exit codes just doing the simple thing here and building then deleting. I think _caching_ builds needs some more thinking and should be a separate PR probably under an opt-in flag to start.

Reply:
The difference and reason for this PR as I understand it is that we can save time linking (what `go build` can optimize for) vs `go run`.

Reply:
Folks we're not setting up this PR  and @dixler  for success by moving the goalposts.

At the point of this morning we're trying to agree on having two changes:

1. https://github.com/pulumi/pulumi/issues/10350 will only use `go build -o $tempFile`, no caching, ship-by-default, feature-flag-to-turn-off, unblock some UX changes

2. https://github.com/pulumi/pulumi/issues/10473 to use `go build -o $cachedFile`, ship-under-flag-first, caching to improve the speed, depends on 10350

Are we good with this? If you need to comment on fixing the design let's take the discussion into those tickets and resolve it there. Then we need to give @dixler time to get the PR lined up.

Perhaps @dixler close this PR and let's start afresh? 🤔